ledger/complete/mtrie/trieCache.go

```diff
@@ -0,0 +1,144 @@
+package mtrie
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+type OnTreeEvictedFunc func(tree *trie.MTrie)
+
+// TrieCache caches tries into memory, it acts as a fifo queue
+// so when it reaches to the capacity it would evict the oldest trie
+// from the cache.
+//
+// Under the hood it uses a circular buffer
+// of mtrie pointers and a map of rootHash to cache index for fast lookup
+type TrieCache struct {
+	tries         []*trie.MTrie
+	lookup        map[ledger.RootHash]int // index to item
+	lock          sync.RWMutex
+	capacity      int
+	tail          int // element index to write to
+	count         int // number of elements (count <= capacity)
+	onTreeEvicted OnTreeEvictedFunc
+}
+
+// NewTrieCache returns a new TrieCache with given capacity.
+func NewTrieCache(capacity uint, onTreeEvicted OnTreeEvictedFunc) *TrieCache {
+	return &TrieCache{
+		tries:         make([]*trie.MTrie, capacity),
+		lookup:        make(map[ledger.RootHash]int, capacity),
+		lock:          sync.RWMutex{},
+		capacity:      int(capacity),
+		tail:          0,
+		count:         0,
+		onTreeEvicted: onTreeEvicted,
+	}
+}
+
+// Purge removes all mtries stored in the buffer
+func (tc *TrieCache) Purge() {
+	tc.lock.Lock()
+	defer tc.lock.Unlock()
+
+	if tc.count == 0 {
+		return
+	}
+
+	toEvict := 0
+	for i := 0; i < tc.capacity; i++ {
+		toEvict = (tc.tail + i) % tc.capacity
+		if tc.onTreeEvicted != nil {
+			if tc.tries[toEvict] != nil {
+				tc.onTreeEvicted(tc.tries[toEvict])
+			}
+		}
+		tc.tries[toEvict] = nil
+	}
+	tc.tail = 0
+	tc.count = 0
+	tc.lookup = make(map[ledger.RootHash]int, tc.capacity)
+}
+
+// Tries returns elements in queue, starting from the oldest element
+// to the newest element.
+func (tc *TrieCache) Tries() []*trie.MTrie {
+	tc.lock.RLock()
+	defer tc.lock.RUnlock()
+
+	if tc.count == 0 {
+		return nil
```

Comment:
suggestion, if you return an empty slice, it would make the job of caller easier. they won't have to check for `nil` anymore

Reply:
hey @taabodim we are going to have a follow up PR for improvements, going to include your suggestions in that PR.